### DISCUSSION TITLE
Is it possible to ignore handshake?

### DISCUSSION THREAD

**Discussion by PseudoPort** [mitmproxy/mitmproxy#5510] [Q&A]
I've been playing with mitmproxy for a while but I still could not get the idea. 
The idea is mitmproxy read Client Hello, then ignore or bypass the next flow, and then give response to client.
Is it possible?

**Comment by meitinger** [comment]
If I understand your question correctly, you'd like to ignore flows based on the client hello?
In that case, you can write an add-on using the `tls_clienthello` hook. This will give you an instance of `ClientHelloData`, which allows you to inspect the `client_hello` and also set `ignore_connection`, which will then ignore the flow.

Here's an example of an existing add-on making use of it: https://github.com/mitmproxy/mitmproxy/blob/a9697398750ea2993047f51708ddd86a746ba7dc/examples/contrib/tls_passthrough.py#L90

If the decision is solely based on the host name, you can also look into the `--ignore-hosts` option.

**Reply by PseudoPort** [reply]
So here is what i wanna gonna do with mitmproxy

ClientHello -> Read value of tls extension -> Bypass next flow and give client a http response

Some snippet of addon i have created:

```python
def tls_clienthello(self, data: tls.ClientHelloData):
    ctx.options.connection_strategy == "lazy"

    for e in data.client_hello.extensions:
       if e[0] == 1000:
          token = e[1]
          ctx.log.info("Token: "+e[1])
          continue

    ## //TODO
    ## Cancel next flow and give client a http response
```
And then i found some snippet to ignore next layer:
```python
   def next_layer(self, nextlayer: proxy.layer.NextLayer):
      is_tls = is_tls_record_magic(nextlayer.data_client())
      if(is_tls):
         nextlayer.layer = layers.TCPLayer(nextlayer.context, ignore=True)
```
next_layer hook ignored tls_clienthello so i cannot get the extension data.

How can i cancel next layer after i get the extension data and give the response back?

**Reply by meitinger** [reply]
So in that case, you don't really want to ignore the flow, you want to alter it, i.e. send back your reply instead of the server's reply, correct?
In that case, use `tls_clienthello` to mark the connection, e.g. `setattr(data.context.client, "has_tls_extension", True)`.
Then use the `request` hook to check for `flow.client_conn.has_tls_extension`. If `True`, create your own response, i.e. `flow.response = Response.make(...)`.

**Reply by rosydawn6** [reply]
For those also having difficulty with the subject of ignoring flows/events I was trying to put this in the def tcp_start lifecycle event and was not seeing any traffic through my tcp addon.  Through the example addon https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/tcp-simple.py I saw that --tcp-hosts ".*" must be specified when launching mitmproxy or no traffic will flow through any custom addons at the tcp life cycle layer.
